core/src/main/java/org/neo4j/graphalgo/core/utils/DisjointSetStruct.java

```diff
@@ -0,0 +1,72 @@
+package org.neo4j.graphalgo.core.utils;
+
+/**
+ * A structure to manage disjoint sets of ints e.g. nodeIds
+ *
+ * @author mknblch
+ */
+public class DisjointSetStruct {
+
+    private final int[] parent;
+    private final int[] depth;
+
+    /**
+     *
+     * @param capacity the capacity (maximum node id)
+     */
+    public DisjointSetStruct(int capacity) {
+        parent = new int[capacity];
+        depth = new int[capacity];
+        for (int i = 0; i < capacity; i++) {
+            parent[i] = i;
+        }
+    }
+
+    /**
+     * find setId of element p.
+     * @param p the element in the set we are looking for
+     * @return an id of the set it belongs to
+     */
+    public int find(int p) {
+        if (parent[p] == p) return p;
+        // path compression optimization
+        parent[p] = find(parent[p]); // balance tree while traversing
```

Comment:
You are true. I've seen 2 possible optimizations in the wild. One is path-halving which works nonrecursive but might at the end not be as efficient as this one. The other one is path compression by recursivly shortening the path during traversion. According to wiki we should get an overall complexity of O(a(n)) ( a(..) is Inverse-Ackermann function). Anyways.. I'd like to do some real-world tests and benchmars on both implementations.